pandas/core/series.py

```diff
@@ -1043,7 +1046,15 @@ def _set_with_engine(self, key, value):
         # fails with AttributeError for IntervalIndex
         loc = self.index._engine.get_loc(key)
         validate_numeric_casting(self.dtype, value)
-        self._values[loc] = value
+        dtype, _ = infer_dtype_from_scalar(value)
+        if is_dtype_equal(self.dtype, dtype):
+            self._values[loc] = value
+        else:
+            # This only raises when index contains tuples
+            try:
+                self.loc[key] = value
```

Comment:
this is a very strange change, why is it needed?

Reply:
Yeah, that is pretty ugly, but a ``loc`` limitation unfortunately. If the index is a regular index where values are tuples, loc does not work.

Example:
```
obj = Series([1, 2], index=[(0,1), (1,2)])
obj[(0,1)] = "test"
```
This runs in the loc case and loc raises
```
KeyError: "None of [Int64Index([0, 1], dtype='int64')] are in the [index]"
```